opt: detect invalid tuple comparison with ANY

RaduBerinde opened the pull request:
Fixes #61229.

Release justification: backportable bug fix.

Release note (bug fix): fixed a case where an invalid tuple comparison using
ANY was causing an internal error; we now return "unsupported
comparison operator".

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61574)
<!-- Reviewable:end -->

mgartner reviewed (commented):
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61574)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/sem/tree/type_check.go, line 1888 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61574#-MVHsxPH4MpbxsYNCUBV:-MVHsxPH4MpbxsYNCUBW:bycpsc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7de01ee1927392dfcd5449d7abc9ec7c3c673d18/pkg/sql/sem/tree/type_check.go#L1888)):*
> ```Go
> 		r := rightType.TupleContents()
> 		if len(l) != len(r) {
> 			return false
> ```

Can you add a regression test for this case too? The behavior seems correct - PG errors with `select (1, 1) = ANY (VALUES (1, 2, 3))` but CRDB 20.2 does not - so it'd be good to validate it.
___
*[pkg/sql/sem/tree/type_check.go, line 1895 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61574#-MVHtLvq1CL14slquNYV:-MVHtLvq1CL14slquNYW:b-a7assh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7de01ee1927392dfcd5449d7abc9ec7c3c673d18/pkg/sql/sem/tree/type_check.go#L1895)):*
> ```Go
> 			}
> 			if !deepCheckValidCmpOp(ops, l[i], r[i]) {
> 				return false
> ```

Might be good to add a test for nested tuples to cover this path too.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: detect invalid tuple comparison with ANY

Fixes #61229.

Release justification: backportable bug fix.

Release note (bug fix): fixed a case where an invalid tuple comparison using
ANY was causing an internal error; we now return "unsupported
comparison operator".

RaduBerinde reviewed (commented):
TFTR, good ideas!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61574#-:-MVI3owk8OZj1vvc38wx:b-ge5bcx)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/type_check.go, line 1888 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61574#-MVHsxPH4MpbxsYNCUBV:-MVI3mnKEn0McgMeFBX_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7de01ee1927392dfcd5449d7abc9ec7c3c673d18/pkg/sql/sem/tree/type_check.go#L1888)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add a regression test for this case too? The behavior seems correct - PG errors with `select (1, 1) = ANY (VALUES (1, 2, 3))` but CRDB 20.2 does not - so it'd be good to validate it.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/type_check.go, line 1895 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61574#-MVHtLvq1CL14slquNYV:-MVI3ngPEvbNm0HPROgf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7de01ee1927392dfcd5449d7abc9ec7c3c673d18/pkg/sql/sem/tree/type_check.go#L1895)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Might be good to add a test for nested tuples to cover this path too.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61574#-:-MVIJQd22kFvsBbmw30X:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

mgartner commented:
Should this be backported to 20.1 as well?
